Native Web Component Support (#58)
salesforce/lwc-rfcs: text/0000-native-web-component-support.md

---

> title: Native Web Component Support

Reviewer: To be more precise can we rename this proposal to Native Custom Element support.

```suggestion
title: Native Custom Element Support
```

status: DRAFTED
created_at: 2021-10-21
updated_at: YYYY-MM-DD
pr: https://github.com/salesforce/lwc-rfcs/pull/58
---

# Native Web Component Support

## Summary

It is possible to author LWC components that can be used natively in an off-platform application by
registering the `CustomElementConstructor` getter.

```js
import { LightningElement } from 'lwc';
class HelloWorld extends LightningElement {}
customElements.define('hello-world', HelloWorld.CustomElementConstructor);
```

```html
<html>
  <body>
    <hello-world></hello-world>
  </body>
</html>
```

> The usage of native web components is generally not allowed on the platform so the main use case is

Reviewer: ```suggestion
The usage of native custom element is generally not allowed on the platform so the main use case is

```

architecture, and as such, it relies on a directory-based module resolution system that is
compatible with how developers create components on the platform. This basically guarantees that no
two parties can define the same component, allowing LWC to side-step a variety of security concerns.

This proposal is about adding native web component support for off-platform users, with the
additional goal of exploring how we might do the same for platform users. Any solution should be
forward-compatible with [Scoped Custom Element Registries].

## Motivation

The long term goal of LWC is to provide a thin abstraction layer on top of native browser APIs.
Adding support for the custom element registry for off-platform applications would be a step towards
that goal.

For platform applications, the major motivation is to facilitate platform integration and allow
developers to migrate existing tooling and solutions without having to utilize LWC.

## Detailed design

### OSS restriction

In order for LWC to support native web components for OSS applications, the following needs to
happen.

1. Prevent LWC components from using the same custom element as a native web component.

> 2. Prevent native web components from using the same custom element as an LWC component.

Reviewer 2: This wording confuses me. Do you mean that LWC and native web components must not collide on names?

3. Remove the restriction that LWC must know about all available components before rendering.

> 

Reviewer 2: Also to solve:
4. Passing attributes vs properties
5. Passing non-primitive HTML types

Reviewer 3: Looking at the test suite for [Custom Elements Everywhere](https://custom-elements-everywhere.com/), it looks like they recommend:

- Arrays/objects should be passed as properties
- Booleans/numbers/strings can be passed as either properties or attributes

In order to pass this test suite, we will also need to handle events with `names-like-this` or `NamesLikeThis`, which LWC currently doesn't support. (At least, in the declarative template syntax, which is what the test suite explicitly tests for.) I.e.:

```html
<!-- currently does not work -->
<some-element onkebab-case={foo}></some-element>
<another-element onPascalCase={foo}></another-element>
```

Reviewer: > In order to pass this test suite, we will also need to handle events with `names-like-this` or `NamesLikeThis`, which LWC currently doesn't support.

For this, we will certainly need to drop `parse5` and use a custom HTML parser. This will be quite a large undertaking.

In order to prevent namespace collisions between LWC components and native web components, LWC
components should be registered in the same custom element registry that native web components will
be registered in.

The third requirement of relaxing the restriction that LWC must know about all available components
before rendering, would be a significant change. Currently, if an unknown component is encountered,
an exception is thrown. This behavior should be modified so that unrecognized custom elements are
rendered as-is with the expectation that they will be natively upgraded. This opens up the
possibility of rendering custom elements that are `HTMLUnknownElement`, which is basically an
extension of `HTMLElement` without any additional fields or methods.

To ensure that only off-platform applications can use this feature, it should be implemented behind
a feature flag (e.g., `ENABLE_CUSTOM_ELEMENT_REGISTRY_RESOLUTION`).

### Platform restriction

In addition to the OSS restrictions outlined above, in order for LWC to support native web

> comopnents for platform applications, the following needs to happen.

Reviewer 2: ```suggestion
components for platform applications, the following needs to happen.

```

> 1. Prevent registration of custom elements in a namespace different from the current one.

Reviewer: This restriction is too aggressive IMO. This will prevent platform developers to reuse existing web component libraries and greatly reduce the usefulness of this proposal.

I also don't think LWC should be in the business of restricting custom elements from being registered based on the current namespace. Locker should be in charge of this.

Reviewer: > I take it you think Locker should hijack the global `customElements.define()` and handle the namespacing there?

Correct

Reviewer 3: Assuming that other frameworks are able to handle custom elements at runtime (props/attrs/events/etc.) without intercepting the `define()` process, my assumption is that LWC should be able to do the same.
